src/effects/backends/builtin/beatgrindeffect.h

```diff
@@ -36,6 +37,7 @@ class BeatGrindState : public EffectState {
     };
 
     mixxx::SampleBuffer delay_buf;
+    std::vector<CSAMPLE> bufferedSample;
```

Comment:
I think this should be a mixxx::SampleBuffer

Reply:
On second thought, I find it confusing that this is a member of the EffectState. I think it would be clearer to just use a pointer local variable in the `process` method.

Reply:
Are you sure it makes more sense? This is pretty much just a place holder for intermediate calculations and not really a sample.
And actually this a reminiscence from the old echo, after getting rid of the send we might not need this at all

Reply:
Yeah, the purpose of the EffectState is to maintain state between cycles of the audio thread. This is only used as an intermediate within the loop; it isn't needed for the next cycle. It isn't even needed at all with the current implementation. Simply copying from the delay buffer to the output buffer would do. But if we add a gater parameter I think an intermediate variable will be needed.

Reply:
but if this is just a plain pointer in the process method, still the memory needs to be allocated by someone, if not state who? 
but yes we might not need this at all. What exactly does the gater parameter do?

Reply:
A plain pointer local variable is on the stack, so that's no problem. Although I think even that is overcomplicated. I think a simple CSAMPLE local variable would do the job.

Reply:
The gater parameter applies a gate to the output:
https://www.youtube.com/watch?v=zxZEsRSXM9U&t=3m
I'm not sure we should combine it with this effect. Maybe that should be its own effect.

Reply:
I am leaning towards leaving the gater out of this effect and making it is own effect.

Reply:
Hmm, you are right. We do need to preserve this value between iterations of the loop, a single variable is enough. 
But still just to understand this, the pointer is allocated in the stack, but where it would point to? The problem here was that the number of channels was not known beforehand so the placeholders needed to be allocated dynamically.
+ 1 for the gate as a separate effect, but I can see the value of it being in the same effect, we can only chain 3 effects right? Or can we chain another chain after one chain?

Reply:
> The problem here was that the number of channels was not known beforehand so the placeholders needed to be allocated dynamically.

The number of channels is known beforehand. The mixxx::EngineParameters passed to the EffectState constructor has that information.

> +1 for the gate as a separate effect, but I can see the value of it being in the same effect, we can only chain 3 effects right? Or can we chain another chain after one chain?

We can chain multiple chains together, but this only works well with Dry/Wet mode. It also doesn't allow saving >3 effects as a single preset, which would be required to use it in the QuickEffect chains.

The more I play with effects, the more I wish I had a way to chain >3 effects. I am starting to think about how this could work for the future. It is tempting to add quick hacks to get the effects we want. But in the long run the effects system will be more flexible if we have each effect do only one thing and let the user chain them in whatever order.

Reply:
Anyway, I don't think this variable is needed at all. Just copy straight from the loop buffer to pOutput.

Reply:
> The more I play with effects, the more I wish I had a way to chain >3 effects. I am starting to think about how this could work for the future.

I could use both my Xone K2s to control every parameter of 6 effects in a chain. :rofl: :rocket:

Reply:
Effects are super fun to play with! 
And a nice way to ruin your mixes too if you overdo it 🤣

Reply:
If we get rid of this variable we need to assign poutput in two places, inside the if when it's reading the loop and in an else statement case write pos < loop length, I think the extra variable is a better design

Reply:
But again I am still confused if the dry/wet mode of chains, we may not need to output the dry signal at all, right?

Reply:
> If we get rid of this variable we need to assign poutput in two places, inside the if when it's reading the loop and in an else statement case write pos < loop length, I think the extra variable is a better design

Sure, either way works.

> But again I am still confused if the dry/wet mode of chains, we may not need to output the dry signal at all, right?

The effect should pass through the dry signal when it isn't outputting the loop. Otherwise there would be a gap of silence.

Reply:
> The effect should pass through the dry signal when it isn't outputting the loop. Otherwise there would be a gap of silence.

Oof, this might not work well with the way Dry+Wet mode is currently implemented. We may need to change how Dry+Wet mode works to get this effect working well in that mode. If I had this effect first in a Dry+Wet mode chain, I would not want it to output the dry signal ever. If it did, that would just increase the volume of the track even though the effect isn't doing anything useful. So it is like this should be an addDryToWet effect sometimes, but only when it isn't outputting from the loop. Only the EffectProcessor would know when this is, so it couldn't be handled by the EngineEffectChain as it is currently. So we could pass a bool parameter down from the EngineEffectChain to the EngineEffect then EffectProcessor indicating whether it is the first "addDryToWet" effect in a Dry+Wet chain. This would require changes to the Echo and Reverb effects too.

Don't worry about this complexity for now. First let's just get it working right in Dry/Wet mode. In this case, the effect simply needs to pass through the dry signal when it isn't outputting from the loop.

Reply:
> we can only chain 3 effects right?

https://github.com/mixxxdj/mixxx/pull/2618#issuecomment-621967476